### PR TITLE
[ir] [autodiff] Initialize ADStack with a zero

### LINKED ISSUE #1781: [bug] [autodiff] Empty AdStack loaded
test case:
`test_ad_fibonacci_index()` in test_ad_for.py
log:
```
[I 08/26/20 19:11:30.900] [compile_to_offloads.cpp:taichi::lang::irpass::`anon
ymous-namespace'::make_pass_printer::<lambda_fe1d620add3df83d4ee306f9c0ab10ca>
::operator ()@18] [fib_c5_0_grad_grad] Simplified I:
kernel {
  <i32 x1> $0 = const [5]
  <i32 x1> $1 = const [0]
  <i32 x1> $2 = const [1]
  <i32 x1> $3 = const [10]
  $4 : for in range($1, $3) (vectorize 1) block_dim=adaptive {
    <i32 x1> $5 = loop $4 index 0
    <f32*x1> $6 = global ptr [S6place_f32], index [$5] activate=true
    <f32 x1> $7 = global load $6
    <f32*x1> $8 = global ptr [S10place_f32], index [] activate=true
    <f32 x1> $9 = atomic add($8, $7)
  }
  $10 : for in range($1, $0) (vectorize 1) block_dim=adaptive {
    <i32 x1> $11 = alloca
    <i32 x1> $12 = alloca
    <i32 x1> $13 : local store [$12 <- $2]
    $14 : for in range($1, $0) (vectorize 1) block_dim=adaptive {
      <i32 x1> $15 = local load [ [$12[0]]]
      <i32 x1> $16 = local load [ [$11[0]]]
      <i32 x1> $17 = add $16 $15
      <i32 x1> $18 : local store [$11 <- $15]
      <i32 x1> $19 : local store [$12 <- $17]
      <f32*x1> $20 = global ptr [S2place_f32], index [$17] activate=true
      <f32 x1> $21 = global load $20
      <f32*x1> $22 = global ptr [S6place_f32], index [$17] activate=true
      <f32 x1> $23 = atomic add($22, $21)
    }
  }
}
[I 08/26/20 19:11:30.903] [compile_to_offloads.cpp:taichi::lang::irpass::`anon
ymous-namespace'::make_pass_printer::<lambda_fe1d620add3df83d4ee306f9c0ab10ca>
::operator ()@18] [fib_c5_0_grad_grad] Gradient:
kernel {
  <i32 x1> $0 = const [5]
  <i32 x1> $1 = const [0]
  <i32 x1> $2 = const [1]
  <i32 x1> $3 = const [10]
  $4 : for in range($1, $3) (vectorize 1) block_dim=adaptive {
    <i32 x1> $5 = loop $4 index 0
    <f32*x1> $6 = global ptr [S6place_f32], index [$5] activate=true
    <f32*x1> $7 = global ptr [S10place_f32], index [] activate=true
    <f32*x1> $8 = global ptr [S12place_f32], index [] activate=true
    <f32 x1> $9 = global load $8
    <f32*x1> $10 = global ptr [S8place_f32], index [$5] activate=true
    <f32 x1> $11 = atomic add($10, $9)
  }
  $12 : for in range($1, $0) (vectorize 1) block_dim=adaptive {
    <f32 x1> $13 = stack alloc (max_size=16)
    <i32 x1> $14 = stack alloc (max_size=16)
    <i32 x1> $15 = stack alloc (max_size=16)
    <i32 x1> $16 = stack alloc (max_size=16)
    <i32 x1> $17 : stack push $16, val = $2
    $18 : for in range($1, $0) (vectorize 1) block_dim=adaptive {
      <i32 x1> $19 = stack load top $16
      <i32 x1> $20 = stack load top $15 // <------------------------------ empty stack!
      <i32 x1> $21 = add $20 $19
      <i32 x1> $22 : stack push $14, val = $21
      <i32 x1> $23 = stack load top $14
      <i32 x1> $24 : stack push $15, val = $19
      <i32 x1> $25 : stack push $16, val = $23
      <f32*x1> $26 = global ptr [S2place_f32], index [$23] activate=true
      <f32 x1> $27 = global load $26
      <f32 x1> $28 : stack push $13, val = $27
      <f32*x1> $29 = global ptr [S6place_f32], index [$23] activate=true
    }
    $30 : reversed for in range($1, $0) (vectorize 1) block_dim=adaptive {
      <i32 x1> $31 = stack load top $14
      <f32*x1> $32 = global ptr [S8place_f32], index [$31] activate=true
      <f32 x1> $33 = global load $32
      <f32 x1> $34 : stack acc adj $13, val = $33
      <f32 x1> $35 = stack load top adj $13
      <f32 x1> $36 : stack pop $13
      <f32*x1> $37 = global ptr [S4place_f32], index [$31] activate=true
      <f32 x1> $38 = atomic add($37, $35)
      <i32 x1> $39 : stack pop $16
      <i32 x1> $40 : stack pop $15
      <i32 x1> $41 : stack pop $14
    }
    <i32 x1> $42 : stack pop $16
  }
}

```

_Originally posted by @xumingkuan in https://github.com/taichi-dev/taichi/pull/1752#issuecomment-680815501_

### PR DESCRIPTION
**Pull request by yuanming-hu**
Related issue = close #1781

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/transforms/auto_diff.cpp [hunk 1]
```diff
@@ -200,9 +200,18 @@ class ReplaceLocalVarWithStacks : public BasicStmtVisitor {
...
-          get_current_program().config.ad_stack_size));
+      auto dtype = alloc->ret_type.data_type;
+      auto stack_alloca = Stmt::make<StackAllocaStmt>(
+          dtype, alloc->get_kernel()->program.config.ad_stack_size);
+      auto stack_alloca_ptr = stack_alloca.get();
+
+      alloc->replace_with(std::move(stack_alloca));
+
+      // Note that unlike AllocStmt, StackAllocaStmt does NOT have an 0 as
+      // initial value. Therefore here we push an initial 0 value.
```
Comment (archibate) [inline taichi/transforms/auto_diff.cpp:211]:
```suggestion
      // Note that unlike AllocaStmt, StackAllocaStmt does NOT have an 0 as
      // initial value. Therefore here we push an initial 0 value.
```
How about enforce `StackAllocaStmt` to be zero initialized over backends like CC did in #1752?

**Review by archibate — APPROVED** [review]
LGTMig!

File: taichi/transforms/auto_diff.cpp (hunk 1 above)
Comment (xumingkuan) [reply taichi/transforms/auto_diff.cpp:211]: There's nothing needed to be done in the backends. The instruction pushing a zero is emitted into the IR and `StackAllocaStmt` only needs to create an empty stack.

**Comment by archibate** [comment]
Merging this for #1752.

**Pull request merged**